### ISSUE TITLE
`npm list --json` doesn't always finish writing to stdout when piped

### ISSUE THREAD

**Issue by gobengo** [npm/npm#17331]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

Today I tried to upgrade some of my servers to node 8 + npm 5. I use ansible's [`npm` module](http://docs.ansible.com/ansible/npm_module.html), and after the upgrade the module would fail to npm install my app. Digging into the stack traces from ansible, I noticed that ansible's python shells out to npm and does `npm list --json`, then parses the result. But it was failing to parse it as valid JSON.

After some more poking, I can sort of reproduce this, and it appears related to the discussion here: https://github.com/nodejs/node/issues/5927 . Which I've run into before. In short, node can't always write > 64kb of stuff to process.stdout *synchronously*.

In the same built repo, I get this in npm 4.2.0 (which is what I use now and has worked fine)
```
$ npm list --json | wc -c
  110648
```

and this with npm 5.0.3
```
$ npm list --json | wc -c
   65536
```

`npm list --json` on its own does write fully valid JSON to my terminal. But when piped somewhere else, it seems to always stop after 64kb. For example, `npm list --json | jq .` fails. Or even `npm list --json | cat` will stop after 64kb.

### supporting information:

 - `npm -v` prints: 5.0.3
 - `node -v` prints: 7.6.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Mac OS and Linux
 - Network issues:
   - Geographic location where npm was run: SF, CA, USA
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by DrewML** [comment]
I'm seeing the same behavior in `npm` version `5.0.3`, with `npm install --json`.

**Supporting Information**
- `npm -v` prints: `5.0.3`
- `node -v`prints: `v8.1.2`
- `npm config get registry` prints:  `https://registry.npmjs.org/`
- Windows, OS X/macOS, or Linux?: Mac OS `10.12.5`

Run in Austin, TX. No container or proxy used.

I'll add that this doesn't happen when I have an existing `node_modules` dir in place, and a `package-lock.json`. Once I delete both of those the issue happens. Might be related to the 64kb number mentioned by @gobengo (would be a much smaller payload passed to `stdout` when the tree is already on disk).

I also verified this _does_ work for me when using `npm` `4.6.1`, and broke at `5.0.0`.

Unrelated: Thanks for all the hard work on npm 5. It's so nice to work with!

**Comment by zetlen** [comment]
Yeah, this is pretty gross. One workaround is to use `pty` to simulate a terminal, so Node doesn't break the pipe. There's this [nice pty binding to node here][1], thanks to @chjj. Using that, I wrote a [workaround gist][2] so that you can get JSON out of npm5 without using the now-deprecated programmatic API. Maybe it could help unblock you until they fix the issue, @DrewML!

Gist pasted below for speed reading.

```js
/**
 * Convenience function to extract and parse JSON from npm5 responses.
 *
 * As of this writing, npm5 doesn't behave when piped to a non-TTY. It truncates output at 64kb.
 * [Until this issue is resolved]{@link https://github.com/npm/npm/issues/17331}, we must 
 * simulate a real TTY to extract full JSON responses from NPM programmatically. This function
 * uses [pty.js]{@link https://github.com/chjj/pty.js} to simulate a TTY, runs any NPM command
 * with the `--json` flag, and returns a Promise for the parsed JSON of the response.
 *
 * @param {string} cmd - The npm command to run, such as `ls` or `info`.
 * @param {string[]} [args] - Any arguments to the npm command to run. `--json` comes automatically.
 * @param {Object} [opts] - Any options to pass to the `pty.js` spawn function.
 *
 * @returns {Promise} - Resolves with the parsed JSON result or rejects with any errors.
 *
 * @example
 * // resolves with a list of all dependencies of the current project
 * require('n5json')('ls').then(({ dependencies }) => Object.keys(dependencies));
 *
 **/
const { spawn } = require('pty.js');
const concat = require('concat-stream');

function tryParse(buf) {
  try {
    return [null, JSON.parse(buf)];
  } catch(e) {
    return [e];
  }
}

module.exports = function n5json(cmd, args=[], opts) {
  const proc = spawn('npm', [cmd, '--json', ...args], opts);
  const done = new Promise((yay, boo) => {
    proc.on('exit', yay);
    proc.on('error', boo);
  });
  const stdout = new Promise((whee) => proc.stdout.pipe(concat((buf) => whee(buf.toString('utf8')))));
  return Promise.all([done, stdout]).then(([code, stdout]) => {
    const succeeded = code === 0;
    const [parseError, result] = tryParse(stdout);
    if (succeeded) {
      if (parseError) {
        throw Error(`n5json: npm succeeded, but could not parse this response: ${stdout}`);
      } else {
        return result;
      }
    } else {
      const err = new Error(`n5json: npm errored with the following response: ${stdout}`);
      if (result) {
        err.originalError = result;
      }
      throw err;
    }
  });
};
```

[1]: https://github.com/chjj
[2]: https://gist.github.com/zetlen/37a2bd17d7008b23b541ea63dab0e221

**Comment by RoryH** [comment]
This issue is effecting our ability to utilise NPM v5 also. 😞

**Comment by accidentaldeveloper** [comment]
I was having this issue with 5.0.4. I upgraded to 5.1.0 this morning and it resolved the bug. I imagine this is connected to https://github.com/npm/npm/pull/17104.

**Comment by gobengo** [comment]
@accidentaldeveloper Hey thanks for pointing that out. I confirmed that upgrading to 5.1.0 fixes all the symptoms I mentioned in the issue above. I will close this ticket.

Props to @zkat for the fix on #17104 !
